### ISSUE TITLE
About the coltran ,I want ask a idiot question?

### ISSUE THREAD

**Issue by meiguoofa** [google-research/google-research#641]
Absolutely ,you did a excellent job! I'm not familiar tf  and in your paper figure 2 show the input image only have grayscale image,so I want to ask that when training ,Is there only a gray image input? Hope you can reply me about this idiot question,Thanks

**Comment by meiguoofa** [comment]
and I'm so curious about how can a Xg(grayscale image) through N layers Row and Col attention can get X spatial down and color down image if there are not color image as input ,where is the color information come form?

**Comment by MechCoder** [comment]
Hi, thanks for your interest in the paper!

When the colorizer is trained, for a given pixel, the inputs are both:

- The grayscale image
- All the previous ground truth pixels as per raster order.

During sampling, each pixel is generated one at a time, and has access to previously generated pixels.
Does that answer your question?

**Comment by MechCoder** [comment]
Yes that is correct. We have code here that can help understand it better (https://github.com/google-research/google-research/blob/master/coltran/models/colorizer.py#L201)

**Comment by meiguoofa** [comment]
Thank you, because I do not fully understand your paper at present, I will read your paper repeatedly, I think your work is very meaningful, and your timely reply is also of great help to me, thank you again

**Comment by meiguoofa** [comment]
Hi, in your paper , there are a description of network structure as below

>  we decompose the task into three simpler sequential subtasks: coarse low resolution autoregressive colorization parallel color and spatial super-resolution

so i want to ask is the parallel color and spatial super-resolution means **parallel** color and **parallel** spatial super-resolution separately?  in other words, this **parallel** that at this position in your paper is the inherent parallel feature of transformer?

**Comment by meiguoofa** [comment]
and more question is the main function of the context representation obtained from the grayscale image is to shift and scale the vector used for attention? Please forgive me for interrupting you again and again with these small problems ^ ^

**Comment by MechCoder** [comment]
>   this parallel that at this position in your paper is the inherent parallel feature of transformer?

In this case, "parallel" refers to the generation of all the pixels in one forward pass. Both the color and spatial upsampler generate all their outputs in one pass. The colorizer, in comparison is autoregressive; it needs a forward pass to generate a given color at a given row and column index.

>  is the main function of the context representation obtained from the grayscale image is to shift and scale the vector used for attention?

The main function is to provide information from the grayscale image to the colorizer. Otherwise, the colorizer becomes unconditional, i.e it will generate random images independent of the grayscale image. Shifting and scaling the attention vectors is one way to do that, there are other ways, for eg see Table 1 in our paper (https://openreview.net/pdf?id=5NA1PinlGFu)

**Comment by meiguoofa** [comment]
Thank you, your explanation is so easy to understand, thank you very much

**Comment by MechCoder** [comment]
Sure, happy to help!

**Comment by meiguoofa** [comment]
Hi in your paper,there is a description about 

> the grayscale encoder encodes each prior channel independently

 the encoder encodes each prior channel  ,but the grayscale image only have one channel,so the encoder encodes grayscale image or an image with three channel?

**Comment by meiguoofa** [comment]
`gray = tf.image.rgb_to_grayscale(inputs)
        z = self.encoder(gray)`
In your code ,  I see you pass a grayscale image to the enocder network, so how does it encodes each prior channel?

**Comment by meiguoofa** [comment]
In the row/column attention of your encoder, it is one row attention layer follow one column attention layer or attention to all rows and then attention to all columns? I wonder if my expression is clear?

**Comment by meiguoofa** [comment]
In your 3.2 AXIAL TRANSFORMER Sampling section the last sentence is 

> The encoder, then recomputes context to generate the next channel.

So after a channel is generated,the grayscale's , the previous output of the grayscale encoder will enter the grayscale encoder again?If like this , as I ask before, how does it encodes each prior channel ?

**Comment by MechCoder** [comment]
> How does it encodes each prior channel?

There is only one channel. Specifically, In our paper, we convert the 3 channel RGB image into a single channel image with 512 colors. We use a 3 bit RGB image, which has total of 8 symbols per channel. This leads to a total of 8**3=512 colors. See the start of Section 4 for more details.
The main idea is that is saves 2x in terms of sampling speed.

Section 3.2 is meant to be as background for the Axial Transformer. For the original Axial Transformer, we indeed need to supply prior channels as input to the encoder, i.e (Gray, Red, Green).

For our paper, we just supply the gray image as input to the encoder and the decoder produces a single channel image.

>  it is one row attention layer follow one column attention layer or attention to all rows and then attention to all columns?

We follow one row attention layer with one column attention layer.

**Comment by meiguoofa** [comment]
Thanks for your patient explanation，I want to talk about my understanding of the network of your paper. I hope you can help me point out my mistakes ,maybe it will take you a lot of time to read this text because my English is not good. And since I am not familiar with tensorflow now, I cannot understand your paper through code for the time being.
When training：
1.The **low resolution grayscale image** enter the grayscale encoder , after n layer row/col attention layers ,  generated a context representation **C** , and the grayscale encoder also has an **auxiliary parallel model** as show in your paper figure2( Is it the linear and softmax layer? and as you answered before the **parallel** here is also means models of all the pixels in one forward pass? ) models a distribution pc hat , and the context  **C** will as a condition to the  outer decoder‘s attetion layers and bias the outer decoder's input 
2. In your paper figure2 ,the ground truth low resolution coarse color image as input (add the **C**) enter the outer decoder ,then after n layer row and masked column attetion and shiftdown ,then the sum of the outer decoder's output  and the **C** as condition to the inner decoder 's conditional attention layers,  and they also add the embeddings that is the eq(4) in your paper as input to the inner decoder ,then after Linear and softmax layer , models the pc distribution,the distrubition after sampling generate a low resolution coarse color image to the color upsamping?
When testing:
3.When training the input of the outer decoder is ground truth coarse color low resolution image.  _**when testing ,what's the input of the Outer Decoder**_  
4.the previous answer you said  

> we just supply the gray image as input to the encoder and the decoder produces a single channel image.

since you just supply the gray image and produces a single channel image, so the grayscale encoder just run once? I run your code and I found the output of the decoder's shape is (1, 64, 64, 1, 512)
![image](https://user-images.githubusercontent.com/22975697/112343312-f762e480-8cfd-11eb-8489-fecf6d37e318.png)
In this picture,the labels is the coarse color low resolution obtained by ground truth?If so , then the same reason,the upsampler color also make cross entropy between the pc hat (generated from ColorUpsampler) and the pc from ground truth and the ps hat (generated from SpatialUpsampler) make  cross_entropy with ground truth?

**Comment by MechCoder** [comment]
> When training the input of the outer decoder is ground truth coarse color low resolution image. when testing ,what's the input of the Outer Decoder?

Ignore the encoder for now, since the grayscale image is present at both during training and evaluation.
Let's say we want to colorize a 2x2 image. As you said, the ground truth is the input to the outer decoder.

|   |  |
| ---      | ---       |
| a |  b      |
| c    | d |


# Training

**Outputs: Outer Decoder** 
(shifted down)
|   |  |
| ---      | ---       |
| 0 |  0      |
| a    | b |

**Inputs: Inner Decoder**
Ground truth is shifted right:

|   |  |
| ---      | ---       |
| 0 |  a      |
| 0   | c |

Net inputs to the inner decoder are (outer decoder output + ground truth shifted right), i.e

|   |  |
| ---      | ---       |
| 0 |  a      |
| a   | b + c |

# Sampling

## First pixel (Row 1, Column 1)

Note that at training, the input location to the first pixel to both decoders are 0

Inputs are initialized to zero, i.e
|   |  |
| ---      | ---       |
| 0 |  0      |
| 0    | 0 |

**Outputs: Outer Decoder** 
(shifted down)
|   |  |
| ---      | ---       |
| 0 |  0      |
| 0    | 0 |

**Inputs: Inner Decoder**
Ground truth shifted right + outputs of the outer decoder which is still zero, i.e
|   |  |
| ---      | ---       |
| 0 |  0      |
| 0    | 0 |

Now, we sample value **a** for the first pixel

## Second pixel (Row 1,  Column 2)
Note that at training, the input location to the inner decoder for the second pixel is **a**

After sampling the first pixel, inputs are now:
|   |  |
| ---      | ---       |
| a |  0      |
| 0    | 0 |

**Outputs: Outer Decoder** 
(shifted down)
|   |  |
| ---      | ---       |
| 0 |  0      |
| a    | 0 |

**Inputs: Inner Decoder**
Ground truth shifted right:
|   |  |
| ---      | ---       |
| 0 |  a    |
| 0    | 0 |

Added to outputs of the outer decoder:
|   |  |
| ---      | ---       |
| 0 |  a  |
| a   | 0 |

The masking ensures that the second row is not attended to. Now, we sample value **b** for the second pixel

## Third pixel (Row 2, Column 1)
Note that at training, the input location to the inner decoder for the third pixel is **a**

After sampling, the second pixel, inputs are now:
|   |  |
| ---      | ---       |
| a |  b      |
| 0    | 0 |

**Outputs: Outer Decoder** 
(shifted down)
|   |  |
| ---      | ---       |
| 0 |  0      |
| a    | b |

**Inputs: Inner Decoder**
Ground truth shifted right:
|   |  |
| ---      | ---       |
| 0 |  a    |
| 0    | b |

Added to outputs of the outer decoder:
|   |  |
| ---      | ---       |
| 0 |  a  |
| a   | b |

Now, we sample value **c** for the third pixel

## Fourth pixel (Row 2, Column 2)
Note that at training, the input location to the inner decoder for the fourth pixel is **b+c**

After sampling, the third pixel, inputs are now:
|   |  |
| ---      | ---       |
| a |  b      |
| c    | 0 |

**Outputs: Outer Decoder** 
(shifted down)
|   |  |
| ---      | ---       |
| 0 |  0      |
| a    | b |

**Inputs: Inner Decoder**
Ground truth shifted right:
|   |  |
| ---      | ---       |
| 0 |  a    |
| 0    | c |

Added to outputs of the outer decoder:
|   |  |
| ---      | ---       |
| 0 |  a  |
| a   | b + c |

Now, we sample value **d** for the final pixel

The sampling strategy is not obvious, and unfortunately it is not easy to explain in a paper.

**Comment by MechCoder** [comment]
> The low resolution grayscale image enter the grayscale encoder , after n layer row/col attention layers , generated a context representation C , and the grayscale encoder also has an auxiliary parallel model as show in your paper figure2( Is it the linear and softmax layer? and as you answered before the parallel here is also means models of all the pixels in one forward pass? ) models a distribution pc hat , and the context C will as a condition to the outer decoder‘s attetion layers and bias the outer decoder's input . The geaysacale'encoders final ouput is a distribution that models 512 colors on one channel?

Yes, correct. It is a linear + softmax layer. See

https://github.com/google-research/google-research/blob/b6754f9b230ca5dabe7011d0d93ad21b09a8641d/coltran/models/colorizer.py#L89

and

https://github.com/google-research/google-research/blob/b6754f9b230ca5dabe7011d0d93ad21b09a8641d/coltran/models/colorizer.py#L151

>In your paper figure2 ,the ground truth low resolution coarse color image as input (add the C) enter the outer decoder ,then after n layer row and masked column attetion and shiftdown ,then the sum of the outer decoder's output and the C as condition to the inner decoder 's conditional attention layers, and they also add the embeddings that is the eq(4) in your paper as input to the inner decoder ,then after Linear and softmax layer , models the pc distribution,the distrubition after sampling generate a low resolution coarse color image to the color upsamping?

Yes.

> since you just supply the gray image and produces a single channel image, so the grayscale encoder just run once? I run your code and I found the output of the decoder's shape is (1, 64, 64, 1, 512). combined with your previous answer, it should be what the following sentence means in your paper , so you used one channel with 512 coarse color instead of 3bit per-channel with rgb channel

Yes, exactly. It predicts a distribution over 512 colors.

> all of your loss function is cross entropy loss in this paper?

Yes, correct.

> In the class ColorUpsampler, the inputs of the call function size is a (B, M, N, 3),but according to your description, shouldn’t it be [B, M, N, 1, 512]? If I read the wrong code line where can i find the convert in your code?

This is a bit subtle. We convert the coarse color back into a RGB image here (so that to visualize the image)

https://github.com/google-research/google-research/blob/b6754f9b230ca5dabe7011d0d93ad21b09a8641d/coltran/models/colorizer.py#L293

In the upsampler, it is converted to a 3-bit RGB image with 8 symbols per channel over here

https://github.com/google-research/google-research/blob/b6754f9b230ca5dabe7011d0d93ad21b09a8641d/coltran/models/upsampler.py#L64

It is then made into a one-hot representation (B, H, W, 3, 8).
Then, each channel (B, H, W, 8) is super-resolved independently into a distribution over 256 symbols, i.e (B, H, W, 256)
This is then concatenated across channels to give (B, H, W, 3, 256)

The coarse color already models dependencies between channels, so this approach worked.

**Comment by meiguoofa** [comment]
OMG, you are really a friendly and patient person. I can no longer express my gratitude to you in words. Your help to me is really great. I wish you a happy life every day.^ ^

**Comment by meiguoofa** [comment]
> > When training the input of the outer decoder is ground truth coarse color low resolution image. when testing ,what's the input of the Outer Decoder?
> 
> Ignore the encoder for now, since the grayscale image is present at both during training and evaluation.
> Let's say we want to colorize a 2x2 image. As you said, the ground truth is the input to the outer decoder. I use a, b, c and d to denote both the values and locations of the first four pixels.
> 
> a	b
> c	d
> # Training
> **Outputs: Outer Decoder**
> (shifted down)
> 
> 0	0
> a	b
> **Inputs: Inner Decoder**
> Ground truth is shifted right:
> 
> 0	a
> 0	c
> Net inputs to the inner decoder are (outer decoder output + ground truth shifted right), i.e
> 
> 0	a
> a	b + c
> # Sampling
> ## First pixel (Row 1, Column 1)
> Note that at training, the input location to the first pixel to both decoders are 0
> 
> Inputs are initialized to zero, i.e
> 
> 0	0
> 0	0
> **Outputs: Outer Decoder**
> (shifted down)
> 
> 0	0
> 0	0
> **Inputs: Inner Decoder**
> Ground truth shifted right + outputs of the outer decoder which is still zero, i.e
> 
> 0	0
> 0	0
> Now, we sample value **a** for the first pixel
> 
> ## Second pixel (Row 1, Column 2)
> Note that at training, the input location to the inner decoder for the second pixel is **a**
> 
> After sampling the first pixel, inputs are now:
> 
> a	0
> 0	0
> **Outputs: Outer Decoder**
> (shifted down)
> 
> 0	0
> a	0
> **Inputs: Inner Decoder**
> Ground truth shifted right:
> 
> 0	a
> 0	0
> Added to outputs of the outer decoder:
> 
> 0	a
> a	0
> The masking ensures that the second row is not attended to. Now, we sample value **b** for the second pixel
> 
> ## Third pixel (Row 2, Column 1)
> Note that at training, the input location to the inner decoder for the third pixel is **a**
> 
> After sampling, the second pixel, inputs are now:
> 
> a	b
> 0	0
> **Outputs: Outer Decoder**
> (shifted down)
> 
> 0	0
> a	b
> **Inputs: Inner Decoder**
> Ground truth shifted right:
> 
> 0	a
> 0	b
> Added to outputs of the outer decoder:
> 
> 0	a
> a	b
> Now, we sample value **c** for the third pixel
> 
> ## Fourth pixel (Row 2, Column 2)
> Note that at training, the input location to the inner decoder for the fourth pixel is **b+c**
> 
> After sampling, the third pixel, inputs are now:
> 
> a	b
> c	0
> **Outputs: Outer Decoder**
> (shifted down)
> 
> 0	0
> a	b
> **Inputs: Inner Decoder**
> Ground truth shifted right:
> 
> 0	a
> 0	c
> Added to outputs of the outer decoder:
> 
> 0	a
> a	b + c
> Now, we sample value **d** for the final pixel
> 
> The sampling strategy is not obvious, and unfortunately it is not easy to explain in a paper.


Hi, when sampling the third pixel , after shifting right  ,the image should be
> 
> 0	a
> 0	0

**Comment by meiguoofa** [comment]
[coltran/models/layers.py L411](https://github.com/google-research/google-research/blob/1aa8e1e08a12ef9dc428e67ee014fcc0931220fe/coltran/models/layers.py#L411)
The annotation should be 
> rows and column  respectively?

**Comment by meiguoofa** [comment]
Hi [coltran/models/layers.py from L659 to L662](https://github.com/MechCoder/google-research/blob/d61d09ebb669542ca91113c22d7323ba34e338d4/coltran/models/layers.py#L659) The parameter **num_encoder_layers** is 1, so there are only one unmask_row and one unmask_col attentionBlock, so in unmask_row layer, your use attention for all rows separately and in unmask_col layer your use attention for all cols separately ?  

> It performs attention over axis k of the tensor x, mixing information along axis k while keeping information along other axes independent. It is straightforward to implement: axial attention over axis k can be implemented by transposing all axes except k to the batch axis, calling standard attention as a subroutine, then undoing the transpose (an alternative is to use the einsum operation available in most deep learning libraries).

and you perform attention over every row and col in this way  described above?

sorry , I should have understood this by looking at the code，but I have spent a whole day still not clearing my thoughts and thank you for your selfless help

**Comment by MechCoder** [comment]
> Hi, when sampling the third pixel , after shifting right ,the i

Indeed, nice catch. I fixed it.

> The annotation should be rows and column respectively?

Yes. [1, num_columns] is row-attention, self-attention is applied to each row independent of the others. I'll fix the documentatio

> The parameter num_encoder_layers is 1

We used 4 blocks in the papers. The configurations used for the experiments in the paper are here (https://github.com/google-research/google-research/blob/master/coltran/configs/colorizer.py#L60).
Yes, we apply row and column attention as you described. The code for the "SelfAttentionND" layer can be made simpler, but they are heavily optimized for TPU's
